File: TeamCode/src/main/java/org/hermitsocialclub/pandemicpanic/telecat/PersistantTelemetry.java
package org.hermitsocialclub.pandemicpanic.telecat;

import org.firstinspires.ftc.robotcore.external.Telemetry;

import java.util.Calendar;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class PersistantTelemetry {

    private final Telemetry t;
    private static final String logSeperator = "==========================";
    private final long logUUID;
    private final boolean log;
    private ConcurrentHashMap<String, String> telemetryData = new ConcurrentHashMap<String, String>();
    private ConcurrentHashMap<String, String> debugData = new ConcurrentHashMap<String, String>();

    public PersistantTelemetry(Telemetry t) {
        this(t, true);
    }

    public PersistantTelemetry(Telemetry t, boolean log) {
        this.t = t;
        this.logUUID = System.currentTimeMillis();
        this.log = log;
        if (log) {
            FileLogger.logToFile("telemetryLog-" + logUUID + ".cast",
                    AsciicastBuilder.buildHeader(logUUID, "Telemetry Log") + "\n");
            FileLogger.clearFile("telemetryLog-latest.cast");
            FileLogger.logToFile("telemetryLog-latest.cast",
                    AsciicastBuilder.buildHeader(logUUID, "Telemetry Log") + "\n");
        }
    }

    /**
     * Persistently sets telemetry data.
     *
     * @param caption The "key" for this data.
     * @param value   What to set the data to.
     */
    public void setData(String caption, Object value) {
        //if(!telemetryData.contains(caption)) {
        telemetryData.put(caption, value.toString());
        //}
        update();
    }

    public void setXData(String key, String formatKey, Object[] values) {
        telemetryData.put(key, String.format(formatKey, values));
        update();
    }

    public void setData(String key, String formatKey, Object... values) {
        telemetryData.put(key, String.format(formatKey, values));
        update();
    }

    public void removeData(String key) {
        telemetryData.remove(key);
        update();
    }

    public void setDebug(String key, Object value) {
        debugData.put(key, value.toString());
        update();
    }

    public void setXDebug(String key, String formatKey, Object[] values) {
        debugData.put(key, String.format(formatKey, values));
        update();
    }

    public void setDebug(String key, String formatKey, Object... values) {
        debugData.put(key, String.format(formatKey, values));
        update();
    }

    private void update() {
        t.clear();
        if (log) {
            Calendar cal = Calendar.getInstance();
            String time = String.format(Locale.CHINESE, "%tT", cal);
            int timeLength = Math.max(logSeperator.length() - time.length(), 0);
            StringBuilder txtLogBuilder = new StringBuilder(
                    logSeperator.substring(0, (int) Math.ceil(timeLength / 2D))
                            + time
                            + logSeperator.substring(0, (int) Math.floor(timeLength / 2D))
                            + "\n");
            StringBuilder logBuilder = new StringBuilder();
            for (Map.Entry<String, String> e : telemetryData.entrySet()) {
                txtLogBuilder.append(e.getKey() + ": " + e.getValue() + "\n");
                logBuilder.append(e.getKey() + ": " + e.getValue() + "\n\u001b[1G");
            }
            for (Map.Entry<String, String> d : debugData.entrySet()) {
                txtLogBuilder.append(d.getKey() + ": " + d.getValue() + "\n");
                logBuilder.append(d.getKey() + ": " + d.getValue() + "\n\u001b[1G");
            }
            String event = AsciicastBuilder.buildEvent(logUUID, cal.getTimeInMillis(), logBuilder.toString()) + "\n";
            FileLogger.logToFile("telemetryLog-" + logUUID + ".cast", event);
            FileLogger.logToFile("telemetryLog-latest.cast", event);
            FileLogger.logToFile("telemetryLog-" + logUUID + ".log", txtLogBuilder.toString());
        }
        for (Map.Entry<String, String> e : telemetryData.entrySet()) {
            t.addData(e.getKey(), e.getValue());
        }
        for (Map.Entry<String, String> d : debugData.entrySet()) {
            t.addData(d.getKey(), d.getValue());
        }
        t.update();
    }

    /**
     * Erases all telemetry data.
     */
    public void clear() {
        telemetryData.clear();
        t.clearAll();
        t.update();
    }

    public Telemetry getOriginalTelemetry() {
        return t;
    }

}

File: TeamCode/src/main/kotlin/org/hermitsocialclub/telecat/PersistantTelemetry.kt
package org.hermitsocialclub.telecat

import org.firstinspires.ftc.robotcore.external.Telemetry
import java.util.*
import kotlin.collections.LinkedHashMap


class PersistantTelemetry @JvmOverloads constructor(val originalTelemetry: Telemetry, val log: Boolean = true) {

    private val logUUID: Long = System.currentTimeMillis()
    private val telemetryData = Collections.synchronizedMap(LinkedHashMap<String, String>())
    private val debugData = Collections.synchronizedMap(LinkedHashMap<String, String>())

    companion object {
        private const val logSeperator = "=========================="
    }

    init {
        if (log) {
            FileLogger.logToFile(
                "telemetryLog-$logUUID.cast", AsciicastBuilder.buildHeader(logUUID, "Telemetry Log")
            )
            FileLogger.clearFile("telemetryLog-latest.cast")
            FileLogger.logToFile(
                "telemetryLog-latest.cast", AsciicastBuilder.buildHeader(logUUID, "Telemetry Log")
            )
        }
    }

    /**
     * Persistently sets telemetry data.
     *
     * @param caption The "key" for this data.
     * @param value   What to set the data to.
     */
    fun setData(caption: String, value: Any) {
        telemetryData[caption] = value.toString()
        update()
    }

    fun setXData(key: String, formatKey: String, values: Array<Any>) {
        telemetryData[key] = String.format(formatKey, *values)
        update()
    }

    fun setData(key: String, formatKey: String, vararg values: Any) {
        telemetryData[key] = String.format(formatKey, *values)
        update()
    }

    fun removeData(key: String) {
        telemetryData.remove(key)
        update()
    }

    fun setDebug(key: String, value: Any) {
        debugData[key] = value.toString()
        update()
    }

    fun setXDebug(key: String, formatKey: String, values: Array<Any>) {
        debugData[key] = String.format(formatKey, *values)
        update()
    }

    fun setDebug(key: String, formatKey: String, vararg values: Any) {
        debugData[key] = String.format(formatKey, *values)
        update()
    }

    private fun update() {
        originalTelemetry.clear()
        if (log) {
            val cal = Calendar.getInstance()
            val time = String.format(Locale.CHINESE, "%tT", cal)
            val timeLength = Math.max(logSeperator.length - time.length, 0)
            val txtLogBuilder = StringBuilder()
            txtLogBuilder
                .append(logSeperator.substring(0, Math.ceil(timeLength / 2.0).toInt()))
                .append(time)
                .append(logSeperator.substring(0, Math.floor(timeLength / 2.0).toInt()))
                .append("\n")
            val logBuilder = StringBuilder()
            for ((key, value) in telemetryData) {
                txtLogBuilder.append("$key: $value")
                logBuilder.append("$key: $value\n\u001b[1G")
            }
            for ((key, value) in debugData) {
                txtLogBuilder.append("$key: $value\n")
                logBuilder.append("$key: $value\n\u001b[1G")
            }
            txtLogBuilder.append("\n")
            val event = AsciicastBuilder.buildEvent(logUUID, cal.timeInMillis, logBuilder.toString())
            FileLogger.logToFile("telemetryLog-$logUUID.cast", event)
            FileLogger.logToFile("telemetryLog-latest.cast", event)
            FileLogger.logToFile("telemetryLog-$logUUID.log", txtLogBuilder.toString())
        }
        for ((key, value) in telemetryData) {
            originalTelemetry.addData(key, value)
        }
        for ((key, value) in debugData) {
            originalTelemetry.addData(key, value)
        }
        originalTelemetry.update()
    }

    /**
     * Erases all telemetry data.
     */
    fun clear() {
        telemetryData.clear()
        originalTelemetry.clearAll()
        originalTelemetry.update()
    }

}